File: src/go/gin/handlers/json.go
package handlers

import (
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
)

// Canonical /json payload. See contracts/rest/canonical-payloads.md.
//
// The previous version returned {id,name,email,timestamp} and called
// time.Now() inside the loop -- 1000 clock reads per request -- which is work
// no other implementation did and which is not what /json measures.
const (
	defaultJSONItems   = 1000
	maxJSONItems       = 10000
	canonicalCreatedAt = "2026-01-01T00:00:00Z"
)

type jsonItem struct {
	ID        int    `json:"id"`
	UUID      string `json:"uuid"`
	Name      string `json:"name"`
	Email     string `json:"email"`
	CreatedAt string `json:"createdAt"`
	IsActive  bool   `json:"isActive"`
}

// newJSONItem builds item i. Content is a pure function of the index: no
// randomness and no wall-clock, so the payload is identical across languages.
func newJSONItem(i int) jsonItem {
	return jsonItem{
		ID:        i,
		UUID:      fmt.Sprintf("00000000-0000-0000-0000-%012d", i),
		Name:      fmt.Sprintf("Item %d", i),
		Email:     fmt.Sprintf("item%d@benchmark.local", i),
		CreatedAt: canonicalCreatedAt,
		IsActive:  i%2 == 0,
	}
}

// itemCount parses ?n=. On a 1 GbE link n=1000 is network-bound at ~734 rps,
// so the serialization ranking is taken at n=100.
func itemCount(raw string) int {
	if raw == "" {
		return defaultJSONItems
	}
	n, err := strconv.Atoi(raw)
	if err != nil || n < 0 {
		return defaultJSONItems
	}
	if n > maxJSONItems {
		return maxJSONItems
	}
	return n
}

func buildItems(n int) []jsonItem {
	items := make([]jsonItem, n)
	for i := 0; i < n; i++ {
		items[i] = newJSONItem(i)
	}
	return items
}

func JsonHandler(c *gin.Context) {
	n := itemCount(c.Query("n"))

	// The envelope timestamp is the only clock-dependent field and is excluded
	// from the parity hash.
	c.JSON(http.StatusOK, gin.H{
		"items":     buildItems(n),
		"count":     n,
		"timestamp": time.Now().UTC().Format(time.RFC3339),
	})
}

File: src/python/django/benchmark/views/json.py
from datetime import datetime, timezone

from rest_framework.decorators import api_view
from rest_framework.response import Response

# Canonical /json payload. See contracts/rest/canonical-payloads.md.
#
# The previous implementation called datetime.utcnow() inside the list
# comprehension -- 1000 clock reads per request -- and returned a shape that
# matched neither the proto nor any other implementation.
DEFAULT_ITEMS = 1000
MAX_ITEMS = 10000
CANONICAL_CREATED_AT = '2026-01-01T00:00:00Z'


def canonical_item(i):
    """Item content is a pure function of the index: no randomness, no clock."""
    return {
        'id': i,
        'uuid': f'00000000-0000-0000-0000-{i:012d}',
        'name': f'Item {i}',
        'email': f'item{i}@benchmark.local',
        'createdAt': CANONICAL_CREATED_AT,
        'isActive': i % 2 == 0,
    }


def item_count(request):
    """
    Parse ?n=. On a 1 GbE link n=1000 is network-bound at ~734 rps, so the
    serialization ranking is taken at n=100.
    """
    raw = request.GET.get('n')
    if raw is None:
        return DEFAULT_ITEMS
    try:
        n = int(raw)
    except (TypeError, ValueError):
        return DEFAULT_ITEMS
    if n < 0:
        return DEFAULT_ITEMS
    return min(n, MAX_ITEMS)


@api_view(['GET'])
def json_endpoint(request):
    n = item_count(request)
    # The envelope timestamp is the only clock-dependent field and is excluded
    # from the parity hash.
    return Response({
        'items': [canonical_item(i) for i in range(n)],
        'count': n,
        'timestamp': datetime.now(timezone.utc).strftime('%Y-%m-%dT%H:%M:%SZ'),
    })
